apps/mark-scan/pat-device-input/src/patinputd.rs

```diff
@@ -0,0 +1,138 @@
+use pin::Pin;
+use std::{
+    io,
+    process::exit,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread,
+    time::Duration,
+};
+use uinput::{
+    event::{keyboard, Keyboard},
+    Device,
+};
+
+use vx_logging::{log, set_app_name, types::EventType, Disposition, EventId};
+
+mod pin;
+
+const UINPUT_PATH: &str = "/dev/uinput";
+const POLL_INTERVAL: Duration = Duration::from_millis(50);
+const APP_NAME: &str = "vx-mark-scan-pat-input-daemon";
+const IS_CONNECTED_PIN: Pin = Pin::new(478);
+const SIGNAL_A_PIN: Pin = Pin::new(481);
+const SIGNAL_B_PIN: Pin = Pin::new(476);
+
+fn send_key(device: &mut Device, key: keyboard::Key) -> Result<(), uinput::Error> {
+    device.click(&key)?;
+    device.synchronize()?;
+    Ok(())
+}
+
+fn create_virtual_device() -> Device {
+    uinput::open(UINPUT_PATH)
+        .unwrap()
+        .name("PAT Input Daemon Virtual Device")
+        .unwrap()
+        .event(Keyboard::All)
+        .unwrap()
+        .create()
+        .unwrap()
+}
+
+fn set_up_pins() -> io::Result<()> {
+    IS_CONNECTED_PIN.set_up()?;
+    SIGNAL_A_PIN.set_up()?;
+    SIGNAL_B_PIN.set_up()?;
+    Ok(())
+}
+
+fn main() {
+    set_app_name(APP_NAME);
+    log!(EventId::ProcessStarted; EventType::SystemAction);
+
+    let running = Arc::new(AtomicBool::new(true));
+
+    if let Err(e) = ctrlc::set_handler({
+        let running = running.clone();
+        move || {
+            running.store(false, Ordering::SeqCst);
+        }
+    }) {
+        log!(
+            event_id: EventId::ErrorSettingSigintHandler,
+            message: e.to_string(),
+            event_type: EventType::SystemStatus
+        );
+    }
+
+    // Create virtual device for keypress events
+    log!(EventId::CreateVirtualUinputDeviceInit; EventType::SystemAction);
+    let mut device = create_virtual_device();
+    // Wait for virtual device to register
+    thread::sleep(Duration::from_secs(1));
+    log!(
+        event_id: EventId::CreateVirtualUinputDeviceComplete,
+        disposition: Disposition::Success,
+        event_type: EventType::SystemAction
+    );
+
+    log!(EventId::ConnectToPatInputInit);
+
+    // Export pins and set direction
+    set_up_pins().unwrap_or_else(|error| {
+        log!(
+            event_id: EventId::ConnectToPatInputComplete,
+            disposition: Disposition::Failure,
+            message: format!("An error occurred during GPIO pin connection: {error}")
+        );
+    });
```

Comment:
When you're not using the value it seems a little weird to use `unwrap_or_else`. Maybe do this?

```suggestion
    if let Err(error) = set_up_pins() {
        log!(
            event_id: EventId::ConnectToPatInputComplete,
            disposition: Disposition::Failure,
            message: format!("An error occurred during GPIO pin connection: {error}")
        );
    }
```

Also, do we want to continue after this?

Reply:
[5171668](https://github.com/votingworks/vxsuite/pull/4569/commits/517166896e8fd00ab8869627f3026539370d59ee)
good catch, thank you